### PR TITLE
Unroll loops

### PR DESCRIPTION
**Pull request by dcousens**
This PR still has a bit of work to do,  ~~as currently this performance has a bit of a readability loss~~.

### CODE REVIEW

**Comment by dcousens** [comment]
Probably good enough to merge.  No benchmark statistics on me,  but I suspect it'll be quite a speed up.

**Comment by jprichardson** [comment]
I'm sure it is a performance enhancement, but for the sake of posterity, could you do a simple benchmark? Something like hashing 100k/1M 32 byte random buffers would at least provide a little insight? Should be an easy script to hack up.

**Comment by dcousens** [comment]
@jprichardson expected nothing less, I'll get round to it today :)

**Comment by dcousens** [comment]
This was actually slower @jprichardson,  but I'm sure I can get it faster.   Will let you know when it is.

This module has some really odd behaviour,  but thankfully its very similar to the SHA* functions in structure anyway so I'm sure it'll come out in the wash.

**Comment by dcousens** [comment]
ping @fanatid

File: lib/ripemd160.js [hunk 1]
```diff
@@ -59,152 +51,174 @@ function bytesToWords (bytes) {
...
-    bytes.push((words[b >>> 5] >>> (24 - b % 32)) & 0xFF)
-  }
-  return bytes
+function swapIntEndian(x) {
+  return (((x << 8) | (x >>> 24)) & 0x00ff00ff) | (((x << 24) | (x >>> 8)) & 0xff00ff00)
 }
 
-function processBlock (H, M, offset) {
+function swapEndian (buffer, length) {
+  length = length || buffer.length
```
Comment (fanatid) [inline lib/ripemd160.js]: excess line

File: lib/ripemd160.js [hunk 2]
```diff
@@ -59,152 +51,174 @@ function bytesToWords (bytes) {
...
-  for (var b = 0; b < words.length * 32; b += 8) {
-    bytes.push((words[b >>> 5] >>> (24 - b % 32)) & 0xFF)
-  }
-  return bytes
+function swapIntEndian(x) {
+  return (((x << 8) | (x >>> 24)) & 0x00ff00ff) | (((x << 24) | (x >>> 8)) & 0xff00ff00)
 }
 
-function processBlock (H, M, offset) {
+function swapEndian (buffer, length) {
```
Comment (fanatid) [inline lib/ripemd160.js]: gosh, `buffer` is not `Buffer`!

File: lib/ripemd160.js [hunk 3]
```diff
@@ -59,152 +51,174 @@ function bytesToWords (bytes) {
...
 
-  // swap endian
-  for (i = 0; i < 5; i++) {
-    // shortcut
-    var H_i = H[i]
+  for (var i = 0; i < m.length; i += 16) {
+    var chunk = m.slice(i, i + 16)
+
+    // swap endian, also fills empty spots w/ 0's
+    swapEndian(chunk, 16)
```
Comment (fanatid) [inline lib/ripemd160.js]: if you call `bytesToWords` after, I think `swapEndian` can be optimized

File: lib/ripemd160.js [hunk 4]
```diff
@@ -59,152 +51,174 @@ function bytesToWords (bytes) {
...
+    j++
+  }
 
-function f1 (x, y, z) {
-  return ((x) ^ (y) ^ (z))
-}
+  // computation
+  while (j < 16) {
+    loopLow(f1(bl, cl, dl))
+    loopHigh(f5(br, cr, dr) + 0x50a28be6)
```
Comment (fanatid) [inline lib/ripemd160.js]: (f5(br, cr, dr) + 0x50a28be6) | 0
and other..

**Comment by fanatid** [comment]
@dcousens I'll try make PR soon (I hope) with benchmark

**Comment by fanatid** [comment]
@dcousens if branch for this PR is not actual, remove it please :)

**Comment by dcousens** [comment]
Deleted :+1:

**Pull request closed without merging**